lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+#include "dynamatic/Transforms/HandshakeOptimization.h"
+#include "dynamatic/Dialect/Handshake/HandshakeCanonicalize.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/LLVM.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <vector>
+using namespace mlir;
+using namespace dynamatic;
+
+namespace {
+
+struct RemoveBranchMUXPairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::MuxOp>(useOwner1) and
```

Comment:
With my change above this becomes `if (isa<handshake::MuxOp>(useOwner1) && useOwner1 == useOwner2)`.